File: src/text.rs
use serde::Serialize;

/// Text defines the two text options (copy and largetext) for an Alfred
/// Item.
///
/// The copy property is the text that is copied to the clipboard when
/// the user pressed CMD-C. The largetype property is the content displayed
/// when the user presses CMD-L.
#[derive(Debug, Clone, Default, PartialEq, Eq, Hash, Serialize)]
pub struct Text {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) copy: Option<String>,

    #[serde(rename = "largetype", skip_serializing_if = "Option::is_none")]
    pub(crate) large_type: Option<String>,
}

#[cfg(test)]
mod tests {

    use serde_json::json;

    use crate::Item;

    #[test]
    fn test_copy() {
        let item = Item::new("Item").copy_text("will be copied");
        let json = serde_json::to_value(&item.text).unwrap();
        let expected = json!({"copy": "will be copied"});
        assert_eq!(json, expected);
    }

    #[test]
    fn test_large_type() {
        let item = Item::new("Item").large_type_text("how big am I");
        let json = serde_json::to_value(&item.text).unwrap();
        let expected = json!({"largetype": "how big am I"});
        assert_eq!(json, expected);
    }
}

File: haskell/src/Alfrusco/Text.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Text type for Alfred item copy and large type content.
module Alfrusco.Text
  ( TextContent (..)
  ) where

import Data.Aeson (ToJSON (..), (.=))
import Data.Aeson qualified as Aeson
import Data.Text (Text)

-- | Represents the text options (copy and largetype) for an Alfred item.
-- The copy property is the text copied to the clipboard with CMD-C.
-- The largetype property is displayed when the user presses CMD-L.
data TextContent = TextContent
  { textCopy :: Maybe Text
  , textLargeType :: Maybe Text
  }
  deriving (Show, Eq)

instance ToJSON TextContent where
  toJSON (TextContent copy largeType) =
    Aeson.object $ concat
      [ maybe [] (\c -> ["copy" .= c]) copy
      , maybe [] (\l -> ["largetype" .= l]) largeType
      ]
